packages/arb-avm-cpp/data_storage/src/arbcore.cpp

```diff
@@ -546,28 +547,37 @@ std::variant<rocksdb::Status, MachineStateKeys> ArbCore::getCheckpointUsingGas(
     marshal_uint256_t(total_gas, key);
     auto key_slice = vecToSlice(key);
     it->SeekForPrev(key_slice);
-    if (!it->Valid()) {
-        if (!it->status().ok()) {
-            return it->status();
+    while (it->Valid()) {
+        if (after_gas) {
+            it->Next();
+            if (!it->status().ok()) {
+                return it->status();
+            }
+            if (!it->Valid()) {
+                return rocksdb::Status::NotFound();
+            }
         }
-        return rocksdb::Status::NotFound();
-    }
-    if (after_gas) {
-        it->Next();
         if (!it->status().ok()) {
             return it->status();
         }
-        if (!it->Valid()) {
-            return rocksdb::Status::NotFound();
+
+        std::vector<unsigned char> saved_value(
+            it->value().data(), it->value().data() + it->value().size());
+        auto variantkeys = extractMachineStateKeys(saved_value);
+
+        if (std::holds_alternative<MachineStateKeys>(variantkeys)) {
+            // Found checkpoint with machine
+            return std::get<MachineStateKeys>(variantkeys);
         }
+
+        // Checkpoint did not contain machine
+        it->Prev();
```

Comment:
This needs to go forwards if `after_gas` is true

Reply:
good catch